File: sdks/node/src/meta-proxy/http-errors.ts
/**
 * HTTP-status -> `AgenticErrorKind` mapping for MetaProxyClient
 * (ADR-0025a §D8's error table is deferred; this pass reuses the same
 * ADR-0024a §D6 status table `MetaLlmClient` uses — `../meta-llm/http-errors.js`
 * — since none of these statuses are Proxy-specific and D8's Proxy-specific
 * `MetaProxyError` shape (`configuredPlane`/`selectedPlane`/`upstreamStatus`
 * fields) is explicitly out of scope this pass).
 */

import { AgenticError } from "../agentic/index.js";

const PRODUCT = "meta-proxy";

function nonEmpty(value: string, fallback: string): string {
  return value.length > 0 ? value : fallback;
}

/** Minimal response shape this mapper needs — satisfied by `fetch`'s `Response`. */
export interface HttpErrorResponse {
  status: number;
  headers: { get(name: string): string | null };
  text(): Promise<string>;
}

export async function mapMetaProxyHttpError(
  response: HttpErrorResponse,
  operation: string,
  requestId: string,
): Promise<AgenticError> {
  const status = response.status;
  const bodyText = await response.text().catch(() => "");
  const fields = { product: PRODUCT, operation, status, requestId };

  switch (status) {
    case 400:
      return new AgenticError("validation", nonEmpty(bodyText, "invalid request"), {
        ...fields,
        retryable: false,
      });
    case 401:
      return new AgenticError(
        "authentication",
        nonEmpty(bodyText, "local Proxy authentication failed"),
        { ...fields, retryable: false },
      );
    case 403:
      return new AgenticError("permission_denied", nonEmpty(bodyText, "permission denied"), {
        ...fields,
        retryable: false,
      });
    case 404:
      return new AgenticError("not_found", nonEmpty(bodyText, "not found"), {
        ...fields,
        retryable: false,
      });
    case 429: {
      const retryAfterHeader = response.headers.get("retry-after");
      const retryAfterMs = retryAfterHeader ? Number(retryAfterHeader) * 1000 : undefined;
      return new AgenticError("rate_limited", nonEmpty(bodyText, "rate limited"), {
        ...fields,
        retryable: true,
        retryAfterMs,
      });
    }
    case 502:
    case 503:
      return new AgenticError("transport", nonEmpty(bodyText, `upstream error ${status}`), {
        ...fields,
        retryable: true,
      });
    default:
      return new AgenticError("protocol", nonEmpty(bodyText, `unexpected status ${status}`), {
        ...fields,
        retryable: false,
      });
  }
}

File: sdks/python/cognitum/meta_proxy/http_errors.py
"""HTTP-status -> ``AgenticErrorKind`` mapping for MetaProxyClient.

ADR-0025a §D8's Proxy-specific ``MetaProxyError`` shape
(``configured_plane``/``selected_plane``/``upstream_status`` fields) is
explicitly out of scope this pass -- this reuses the same ADR-0024a §D6
status table :mod:`cognitum.meta_llm.http_errors` uses, since none of these
statuses are Proxy-specific.
"""

from __future__ import annotations

from typing import Any

import httpx

from cognitum.agentic import AgenticError

_PRODUCT = "meta-proxy"


def map_meta_proxy_http_error(
    response: httpx.Response, operation: str, request_id: str
) -> AgenticError:
    status = response.status_code
    body_text = response.text
    common: dict[str, Any] = {
        "product": _PRODUCT,
        "operation": operation,
        "status": status,
        "request_id": request_id,
    }

    if status == 400:
        return AgenticError(
            "validation", body_text or "invalid request", retryable=False, **common
        )
    if status == 401:
        return AgenticError(
            "authentication",
            body_text or "local Proxy authentication failed",
            retryable=False,
            **common,
        )
    if status == 403:
        return AgenticError(
            "permission_denied", body_text or "permission denied", retryable=False, **common
        )
    if status == 404:
        return AgenticError("not_found", body_text or "not found", retryable=False, **common)
    if status == 429:
        retry_after_header = response.headers.get("retry-after")
        retry_after_ms = int(float(retry_after_header) * 1000) if retry_after_header else None
        return AgenticError(
            "rate_limited",
            body_text or "rate limited",
            retryable=True,
            retry_after_ms=retry_after_ms,
            **common,
        )
    if status in (502, 503):
        return AgenticError(
            "transport", body_text or f"upstream error {status}", retryable=True, **common
        )
    return AgenticError(
        "protocol", body_text or f"unexpected status {status}", retryable=False, **common
    )


__all__ = ["map_meta_proxy_http_error"]
